lib/event_store_client/adapters/grpc/shared/options/stream_options.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+module EventStoreClient
+  module GRPC
+    module Shared
+      module Options
+        class StreamOptions
+          attr_reader :stream_name, :options
+          private :stream_name, :options
+
+          # @param stream_name [String]
+          # @param options [Hash]
+          # @option options [Integer, Symbol] :from_revision. If number is provided - it is threaded
+          #   as starting revision number. Alternatively you can provide :start or :end value to
+          #   define a stream revision. **Use this option when stream name is a normal stream name**
+          # @option options [Hash, Symbol] :from_position. If hash is provided - you should supply
+          #   it with :commit_position and/or :prepare_position keys. Alternatively you can provide
+          #   :start or :end value to define a stream position. **Use this option when stream name
+          #   is "$all"**
+          def initialize(stream_name, options)
+            @stream_name = stream_name
+            @options = options
+          end
+
+          # @return [Hash]
+          def request_options
+            stream_name == "$all" ? all_stream : stream
+          end
+
+          private
+
+          # @return [Hash]
+          #   Examples:
+          #   ```ruby
+          #   { all: { start: EventStore::Client::Empty.new } }
+          #   ```
+          #   ```ruby
+          #   { all: { end: EventStore::Client::Empty.new } }
+          #   ```
+          #   ```ruby
+          #   { all: { position: { commit_position: 1, prepare_position: 1 } } }
+          #   ```
+          def all_stream
+            position_opt =
+              case options[:from_position]
+              when :start, :end
+                { options[:from_position] => EventStore::Client::Empty.new }
+              when Hash
+                { position: options[:from_position] }
```

Comment:
We need some check here in case like somebody provide options like { from_position :nil } or { from_position: 'xyz' }
Why not go further and implement options with some class with the check of value correctness? For example with dry schema or plain ruby value object class?

Reply:
Classes, generated by google-protobuf already has this feature. So, if you prepare request options in a hash and then pass it to the proper params class, e.g. `EventStore::Client::Streams::ReadReq::Options`, it will raise if there is an invalid value. 

Even more - it correctly handles such definitions, as `oneof`. For example, if you have this in generated file:

```ruby
oneof :stream_option do
  optional :stream_identifier, :message, 1, "event_store.client.StreamIdentifier"
  optional :all, :message, 5, "event_store.client.Empty"
end
```
and you will try to assign both options - only the last assigned option will be kept.